src/crate/client/test_http.py

```diff
@@ -472,79 +495,113 @@ class TestingHTTPServer(HTTPServer):
-        cls(server_address, RetryRequestHandler).serve_forever()
-
+    def run_server(cls, server_address, request_handler_cls):
+        cls(server_address, request_handler_cls).serve_forever()
 
-class RetryOnTimeoutServerTest(TestCase):
 
-    server_address = ("127.0.0.1", 65535)
+class TestingHttpServerTestCase(TestCase):
 
     def __init__(self, *args, **kwargs):
-        super(RetryOnTimeoutServerTest, self).__init__(*args, **kwargs)
-        self.server_process = Process(target=TestingHTTPServer.run_server, args=(self.server_address,))
+        super().__init__(*args, **kwargs)
+        self.assertIsNotNone(self.request_handler)
+        self.server_address = ('127.0.0.1', random.randint(65000, 65535))
+        self.server_process = Process(target=TestingHTTPServer.run_server,
+                                      args=(self.server_address, self.request_handler))
 
     def setUp(self):
-        self.client = Client(["%s:%d" % self.server_address], timeout=5)
         self.server_process.start()
-        time.sleep(.10)
+        self.wait_for_server()
+
+    def wait_for_server(self):
+        print('')
+        while True:
+            print('Waiting for ', self.server_address)
+            try:
+                with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+                    s.connect(self.server_address)
+            except:
+                time.sleep(.25)
+            else:
+                print('Connected to', self.server_address)
+                break
 
     def tearDown(self):
         self.server_process.terminate()
+
+    def clientWithKwargs(self, **kwargs):
+        return Client(["%s:%d" % self.server_address], timeout=1, **kwargs)
+
+
+class RetryOnTimeoutServerTest(TestingHttpServerTestCase):
+
+    request_handler = TimeoutRequestHandler
+
+    def setUp(self):
+        super().setUp()
+        self.client = self.clientWithKwargs()
+
+    def tearDown(self):
+        super().tearDown()
         self.client.close()
 
     def test_no_retry_on_read_timeout(self):
         try:
             self.client.sql("select * from fake")
-        except ConnectionError:
-            pass
+        except ConnectionError as e:
+            self.assertTrue('Read timed out' in e.message,
+                            msg='Error message must contain: Read timed out')
         self.assertEqual(TestingHTTPServer.SHARED['count'], 1)
 
 
-class TestUsernameSentAsHeader(TestCase):
+class TestDefaultSchemaHeader(TestingHttpServerTestCase):
 
-    server_address = ("127.0.0.1", 62535)
+    request_handler = SharedStateRequestHandler
+
+    def setUp(self):
+        super().setUp()
+        self.client = self.clientWithKwargs(schema='my_custom_schema')
+
+    def tearDown(self):
+        super().tearDown()
+        self.client.close()
+
+    def test_default_schema(self):
+        self.client.sql('SELECT 1')
+        self.assertEqual(TestingHTTPServer.SHARED['schema'], 'my_custom_schema')
```

Comment:
Should we do an additional test here to see that queries really use the custom schema?

Reply:
That functionality is tested within CrateDB already

Reply:
So there is a client test in CrateDB which checks that the client passes on the schema correctly?

Reply:
https://github.com/crate/crate/blob/master/sql/src/test/java/io/crate/integrationtests/RestSQLActionIntegrationTest.java#L71-L83

Reply:
Cool!